### PR TITLE
reorganized code, better subcommands api, added region, context and credentials 

### PR DESCRIPTION
**Pull request by juantascon**
- environments are now called deployments
- removed CLIModules in favor of preset Modules
- `__init__.py` files now contain very few LOCs
- region functionality moved to sdk
- username and password go now to a separate config file (credentials)
- better subcommands api including context class
- `get_parser()` functions aren't @classmethods anymore
- added more docs including features and how to extend
- unit tests are now written with the pytest framework, easier to understand and extend

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,10 +77,47 @@ use the following command:
...
-General documentation is listed here:
+More info:
 https://packaging.python.org/distributing/
+
+
+## How to add new commands:
+
+1. Add a subdirectory in alertlogiccli/commands/
+
+2. Write at least 1 subcommand, they must be classes that inherit from Command
```
Comment (anton-b) [inline README.md]: maybe it's worth adding an `examples` directory and put a working skeleton of the command module.

File: share/README.md [hunk 2]
```diff
@@ -18,76 +18,90 @@ The easiest way to install alertlogic-cli is using pip:
...
+
+Example config:
+
+``` ini
+[default]
+api_endpoint = us
+account_id = 123089
+```
+
+Example credentials:
```
Comment (anton-b) [inline share/README.md]: maybe add working credentials and config examples to the examples directory?

**Comment by anton-b** [comment]
Couple questions
* Did you fix unit tests added new ones?
* Can you please make a followup userstory to add Jenkins job(e.g. in the https://hub.openshift.com/quickstarts/98-jenkins-server) hook to run tests and maybe PEP 8 check over the code?

**Comment by juantascon** [comment]
Answers to your questions:

1. I just added a commit were I reworked all unit tests, please take a look
2. done

**Comment by anton-b** [comment]
I've tried to run tests, it didn't work(missing pytest2). There is no such package pytest2 in pip. Please fix and test in VM with CentOS and python 2x

**Comment by juantascon** [comment]
Yes it's only `pytest`, some distros rename it as `pytest2` but I think `pytest` alone is generally more common, I changed it to `pytest`
That error you are getting is because you need to install `mock` lib too

**Comment by anton-b** [comment]
@juantascon That is secondary problem. The main issues i am pointing here out are:
* pytest2 is not found in the pypi(pip) which is standard for any platform
* even if i change to pytest in the make file pytest does not accept wildcards e.g. *_tests

**Comment by anton-b** [comment]
@juantascon ok, take my prev comment back. 

Though, ran test on the centos 7 docker container and i see this:
```
 make test
pytest tests/*_tests.py
======================================================== test session starts ========================================================
platform linux2 -- Python 2.7.5, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
rootdir: /alertlogic-cli, inifile:
collected 14 items

tests/auth_tests.py ...
tests/deployment_mode_tests.py FFFFFF
tests/deployment_resources_tests.py F
tests/deployment_scan_queue_tests.py F
tests/deployment_status_tests.py F
tests/dynapi_tests.py ..

============================================================= FAILURES ==============================================================
____________________________________________________ TestGetMode.test_not_found _____________________________________________________

args = (<tests.deployment_mode_tests.TestGetMode instance at 0x1fa8908>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_mode_tests.TestGetMode instance at 0x1fa8908>

    @httpretty.activate
    def test_not_found(self):
        args = mock_tools.make_args()
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/sources/v1/{account_id}/sources/{deployment_id}".format(**vars(args)),
                               status=404)
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_mode_tests.py:15: TypeError
________________________________________________________ TestGetMode.test_ok ________________________________________________________

args = (<tests.deployment_mode_tests.TestGetMode instance at 0x24505a8>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_mode_tests.TestGetMode instance at 0x24505a8>

    @httpretty.activate
    def test_ok(self):
        args = mock_tools.make_args()
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/sources/v1/{account_id}/sources/{deployment_id}".format(**vars(args)),
                               body='{"source": {"type": "environment", "config": {"deployment_mode": "readonly"}}}',
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_mode_tests.py:47: TypeError
_________________________________________________ TestGetMode.test_invalid_response _________________________________________________

args = (<tests.deployment_mode_tests.TestGetMode instance at 0x21690e0>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_mode_tests.TestGetMode instance at 0x21690e0>

    @httpretty.activate
    def test_invalid_response(self):
        args = mock_tools.make_args()
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/sources/v1/{account_id}/sources/{deployment_id}".format(**vars(args)),
                               body='INVALID_BODY',
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_mode_tests.py:35: TypeError
___________________________________________________ TestGetMode.test_server_error ___________________________________________________

args = (<tests.deployment_mode_tests.TestGetMode instance at 0x21f4950>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_mode_tests.TestGetMode instance at 0x21f4950>

    @httpretty.activate
    def test_server_error(self):
        args = mock_tools.make_args()
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/sources/v1/{account_id}/sources/{deployment_id}".format(**vars(args)),
                               status=500)
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_mode_tests.py:25: TypeError
___________________________________________________ TestSetMode.test_server_fail ____________________________________________________

args = (<tests.deployment_mode_tests.TestSetMode instance at 0x244ab90>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_mode_tests.TestSetMode instance at 0x244ab90>

    @httpretty.activate
    def test_server_fail(self):
        args = mock_tools.make_args(additional_args={"mode": "readonly"})
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/sources/v1/{account_id}/sources/{deployment_id}".format(**vars(args)),
                               body='{"source": {"type": "environment"}}',
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_mode_tests.py:75: TypeError
________________________________________________________ TestSetMode.test_ok ________________________________________________________

args = (<tests.deployment_mode_tests.TestSetMode instance at 0x2451680>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_mode_tests.TestSetMode instance at 0x2451680>

    @httpretty.activate
    def test_ok(self):
        args = mock_tools.make_args(additional_args={"mode": "readonly"})
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/sources/v1/{account_id}/sources/{deployment_id}".format(**vars(args)),
                               body='{"source": {"type": "environment"}}',
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_mode_tests.py:60: TypeError
_____________________________________________________ TestListDeployed.test_ok ______________________________________________________

args = (<tests.deployment_resources_tests.TestListDeployed instance at 0x244e560>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_resources_tests.TestListDeployed instance at 0x244e560>

    @httpretty.activate
    def test_ok(self):
        args = mock_tools.make_args()
        body = json.dumps(LIST_RESOURCES_REPLY_JSON)
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/launcher/v1/{account_id}/{deployment_id}/resources".format(**vars(args)),
                               body=body,
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_resources_tests.py:50: TypeError
____________________________________________________ TestListScanQueues.test_ok _____________________________________________________

args = (<tests.deployment_scan_queue_tests.TestListScanQueues instance at 0x1cf2638>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_scan_queue_tests.TestListScanQueues instance at 0x1cf2638>

    @httpretty.activate
    def test_ok(self):
        args = mock_tools.make_args()
        body = json.dumps(LIST_REPLY_JSON)
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/scheduler/v1/{account_id}/{deployment_id}/list".format(**vars(args)),
                               body=body,
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_scan_queue_tests.py:167: TypeError
_______________________________________________________ TestGetStatus.test_ok _______________________________________________________

args = (<tests.deployment_status_tests.TestGetStatus instance at 0x1a215f0>,), kw = {}

    @functools.wraps(test)
    def wrapper(*args, **kw):
        with httprettized():
>           return test(*args, **kw)

/usr/lib/python2.7/site-packages/httpretty/core.py:1146:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.deployment_status_tests.TestGetStatus instance at 0x1a215f0>

    @httpretty.activate
    def test_ok(self):
        args = mock_tools.make_args()
        body = json.dumps(GET_STATUS_REPLY_JSON)
        httpretty.register_uri(httpretty.GET,
>                              "{api_endpoint}/launcher/v1/{account_id}/environments/{deployment_id}".format(**vars(args)),
                               body=body,
                               status=200,
                               content_type="text/json")
E       TypeError: vars() argument must have __dict__ attribute

tests/deployment_status_tests.py:75: TypeError
================================================ 9 failed, 5 passed in 0.42 seconds =================================================
make: *** [test] Error 1
```

**Comment by anton-b** [comment]
Fyi what i did from the container start:
```
yum install git python
    5  pip
    6  yum install wget -y
    7  wget https://bootstrap.pypa.io/get-pip.py
    8  python get-pip.py
    9  ssh-keygen
   10  less ~/.ssh/id_rsa.pub
   11  git clone git@github.com:alertlogic/alertlogic-cli.git
   12  cd alertlogic-cli/
   13  ls
   14  make test
   15  yum install autoconf
   16  make
   17  yum install automake
   18  make
   19  yum provides make
   20  yum install make
   21  yum provides make
   22  make
   23  make test
   24  git remote add juan git@github.com:juantascon/alertlogic-cli.git
   25  gti fetch jua
   26  gti fetch juan
   27  git fetch juan
   28  git checkout reorg
   29  make
   30  make test
   31  pip install pytest
   32  make test
   33  pip install mock
   34  make test
   35  pip install httppretty
   36  pip install httpretty
   37  make test
   38  pip install requests
   39  make test
   40  vi Makefile
   41  ршыещкн
   42  history
```

**Comment by juantascon** [comment]
Seems like using `vars()` on a `namedtuple` is only possible on newer versions, older versions however should use `_todict()`
I changed the code so we use `argparse.Namespace` instead, I tested it in centos7 and it works there too

**Comment by anton-b** [comment]
test

**Comment by anton-b** [comment]
test

**Pull request merged**